### ISSUE TITLE
Enable Multi-GPU Training (DDP)

### ISSUE THREAD

**Issue by hmehryar** [hmehryar/Hm.JetscapeMl#351]
### Summary
Convert our single-GPU training pipeline into a fully distributed, multi-GPU setup using PyTorch’s `DistributedDataParallel` (DDP). This covers changes to:

- `loader.py`  
- `train_mamba.py` (or your main training script/notebook)  
- SLURM launch script  
- Environment setup  

---

### 📋 Checklist

- [ ] **Environment**  
  - [ ] Install `importlib_metadata` (for `torchrun` support)  
    ```bash
    pip install importlib_metadata
    # or
    conda install -c conda-forge importlib_metadata
    ```
- [ ] **SLURM script** (`.sbatch`)  
  - [ ] Set `--ntasks-per-node` to number of GPUs  
  - [ ] Reserve GPUs via `--gres=gpu:<type>:<N>`  
  - [ ] Launch with `srun python -m torch.distributed.run …` or `torchrun --nproc_per_node=$NGPUS`  

- [ ] **`loader.py`**  
  - [ ] Import `DistributedSampler`  
  - [ ] Replace `shuffle=True` DataLoaders with `DistributedSampler`  
  - [ ] Return samplers alongside loaders  

- [ ] **`train_mamba.py`** (or your main script)  
  - **Initialization (top of file)**  
    - [ ] `import torch.distributed as dist`  
    - [ ] `dist.init_process_group(backend="nccl")`  
    - [ ] Read `LOCAL_RANK`, set `torch.cuda.set_device(local_rank)` and `device = torch.device("cuda", local_rank)`  
  - **Model definition**  
    - [ ] Move model to `device`  
    - [ ] Wrap with `torch.nn.parallel.DistributedDataParallel`  
  - **DataLoaders**  
    - [ ] Unpack samplers from `get_dataloaders`  
    - [ ] Call `train_sampler.set_epoch(epoch)` at each epoch start  
  - **Checkpointing & Logging**  
    - [ ] Guard prints and saves with `if local_rank == 0:`  
  - [ ] **Teardown**  
    - [ ] `dist.destroy_process_group()` after training  

---

### 📓 Example SLURM Launch

```bash
#!/bin/bash
#SBATCH --job-name=ml-jet-ddp
#SBATCH --partition=ecscp
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --gres=gpu:nvidia_a100_80gb_pcie_1g.10gb:4
#SBATCH --cpus-per-task=4
#SBATCH --mem=8G
#SBATCH --time=504:00:00
#SBATCH --qos=express

module load cuda/11.7
source activate pytorch

srun python -m torch.distributed.run \
     --nproc_per_node=$SLURM_NTASKS_PER_NODE \
     train_mamba.py \
     --epochs 50 \
     --batch-size 128 | tee output.log

**Comment by hmehryar** [comment]
> ⚠️ **Note:** Unfortunately, deploying a multi-GPU DDP job on the A100 nodes under the **express** queue isn’t possible right now. Those nodes are partitioned with MIG slices, and without admin privileges we cannot reconfigure the MIG setup to expose full GPUs. As a result, only single-GPU jobs can run in this queue.
